### PR TITLE
[dataset] add 'Info' and 'Components' types

### PR DESCRIPTION
**Pull request by abtink**
This commit adds `Dataset::Info` and `Dataset::Components` types
mirroring `otOperationalDataset` & `otOperationalDatasetComponents`
definitions and providing helper methods to get/set different fields
in the dataset.

### CODE REVIEW

File: src/core/meshcop/dataset.hpp [hunk 1]
```diff
@@ -73,6 +75,528 @@ class Dataset
...
+         *
+         */
+        void SetChannelMask(otChannelMask aChannelMask)
+        {
+            mChannelMask                      = aChannelMask;
+            mComponents.mIsChannelMaskPresent = true;
+        }
+
+        /**
+         * This method populates the Dataset withr given and/or random fields.
```
Comment (Irving-cl) [inline src/core/meshcop/dataset.hpp:584]:
```suggestion
         * This method populates the Dataset with given and/or random fields.
```

File: src/core/meshcop/dataset.hpp [hunk 2]
```diff
@@ -73,6 +75,528 @@ class Dataset
...
+         *
+         * The Channel, Channel Mask, and Security Policy Flags are given as input parameters. The Mater Key, PSKc,
```
Comment (Irving-cl) [inline src/core/meshcop/dataset.hpp:586]:
```suggestion
         * The Channel, Channel Mask, and Security Policy Flags are given as input parameters. The Master Key, PSKc,
```

File: src/core/meshcop/dataset.hpp [hunk 3]
```diff
@@ -73,6 +75,528 @@ class Dataset
...
+         * to "OpenThread-%04x" with PAN ID appended as hex.
+         *
+         * @param[in] aChannel          The Channel value to use in the Dataset.
+         * @param[in] aChannelMask      The Channel Mask value to use in the Dataset.
+         * @param[in] aSecurityFlags    The Security Policy Flags.
+         *
+         * @retval OT_ERROR_NONE If the Dataset was generated successfully.
+         *
+         */
+        otError GenerateRandom(uint16_t aChannel, otChannelMask aChannelMask, uint8_t aSecurityFalgs);
```
Comment (Irving-cl) [inline src/core/meshcop/dataset.hpp:597]:
```suggestion
        otError GenerateRandom(uint16_t aChannel, otChannelMask aChannelMask, uint8_t aSecurityFlags);
```

**Review by Irving-cl — APPROVED** [review]
LGTM.

Some typos.

File: src/core/meshcop/dataset.cpp [hunk 4]
```diff
@@ -51,6 +51,40 @@ namespace MeshCoP {
 using ot::Encoding::BigEndian::HostSwap16;
 using ot::Encoding::BigEndian::HostSwap32;
 
+otError Dataset::Info::GenerateRandom(uint16_t aChannel, otChannelMask aChannelMask, uint8_t aSecurityFalgs)
```
Comment (Irving-cl) [inline src/core/meshcop/dataset.cpp:54]:
```suggestion
otError Dataset::Info::GenerateRandom(uint16_t aChannel, otChannelMask aChannelMask, uint8_t aSecurityFlags)
```

File: src/core/meshcop/dataset.cpp [hunk 5]
```diff
@@ -51,6 +51,40 @@ namespace MeshCoP {
...
+{
+    otError error;
+
+    Clear();
+
+    mActiveTimestamp       = 1;
+    mChannel               = aChannel;
+    mChannelMask           = aChannelMask;
+    mSecurityPolicy.mFlags = aSecurityFalgs;
```
Comment (Irving-cl) [inline src/core/meshcop/dataset.cpp:63]:
```suggestion
    mSecurityPolicy.mFlags = aSecurityFlags;
```

**Review by Irving-cl — APPROVED** [review]

File: src/core/meshcop/dataset.hpp (hunk 1 above)
Comment (abtink) [reply src/core/meshcop/dataset.hpp]: Thanks @Irving-cl

File: src/core/meshcop/dataset_manager_ftd.cpp [hunk 6]
```diff
@@ -333,26 +320,8 @@ otError ActiveDataset::CreateNewNetwork(otOperationalDataset &aDataset)
...
-    aDataset.mComponents.mIsMeshLocalPrefixPresent = true;
-    aDataset.mComponents.mIsPanIdPresent           = true;
-    aDataset.mComponents.mIsChannelPresent         = true;
-    aDataset.mComponents.mIsPskcPresent            = true;
-    aDataset.mComponents.mIsSecurityPolicyPresent  = true;
-    aDataset.mComponents.mIsChannelMaskPresent     = true;
-
-exit:
-    return error;
+    return aDataset.GenerateRandom(preferredChannels.ChooseRandomChannel(), supportedChannels.GetMask(),
```
Comment (jwhui) [inline src/core/meshcop/dataset_manager_ftd.cpp:323]: What about passing `preferredChannels` and let `GenerateRandom()` select a random channel?

Comment (abtink) [reply src/core/meshcop/dataset_manager_ftd.cpp]: In the new push, I changed it to pass `aInstance` and then have the `Dataset::Info::GenerateRandom()` method itself read the preferred/supported masks and decide on channel.

**Review by jwhui — APPROVED** [review]

**Pull request merged**